multi/grlib-multi/spacewire.c

```diff
@@ -0,0 +1,829 @@
+		(void)munmap((void *)dev->txDesc, descSz);
+	}
+}
+
+
+static size_t spw_txMsgToPacket(const uint8_t *buf, spw_txPacket_t *packet)
+{
+	packet->flags = (buf[0] << 24) | (buf[1] << 16) | (buf[2] << 8) | buf[3];
+	packet->dataLen = (buf[4] << 24) | (buf[5] << 16) | (buf[6] << 8) | buf[7];
+
+	packet->hdr = &buf[SPW_TX_MIN_BUFSZ];
+	packet->data = &buf[SPW_TX_MIN_BUFSZ] + (packet->flags & 0xff);
+
+	return SPW_TX_MIN_BUFSZ + (packet->flags & 0xffu) + packet->dataLen;
+}
+
+
+static size_t spw_rxPacketToMsg(const uint32_t flags, const size_t rxLen, const uint8_t *rx, uint8_t *buf)
+{
+	buf[0] = flags >> 24;
+	buf[1] = (flags >> 16) & 0xffu;
+	buf[2] = (flags >> 8) & 0xffu;
+	buf[3] = flags & 0xffu;
+
+	memcpy(&buf[SPW_RX_MIN_BUFSZ], rx, rxLen);
+
+	return SPW_RX_MIN_BUFSZ + rxLen;
+}
+
+
+/* Interrupt handling */
+
+
+static int spw_irqHandler(unsigned int n, void *arg)
+{
+	(void)n;
+
+	spw_dev_t *dev = (spw_dev_t *)arg;
+
+	/* TX IRQ */
+	while (dev->sentDesc != dev->lastTxDesc) {
+		if (dev->txDesc[dev->sentDesc].ctrl & TX_DESC_EN) {
+			/* Not sent yet */
+			break;
+		}
+
+		dev->sentDesc = (dev->sentDesc + 1) % SPW_TX_DESC_CNT;
+		dev->txDescFree++;
+	}
+
+	/* For RX IRQ only cond in kernel is signalled */
+
+	return 1;
+}
+
+
+/* Operations on device */
+
+
+/* Blocks execution until all packets are transmitted. */
+static int spw_transmitWait(spw_dev_t *dev, const uint8_t *buf, const size_t nPackets)
+{
+	if (nPackets > SPW_TX_DESC_CNT) {
+		return -EINVAL;
+	}
+
+	(void)mutexLock(dev->txLock);
+
+	TRACE("nPackets: %d", nPackets);
+
+	/* Setup descriptors */
+	size_t firstDesc = dev->lastTxDesc;
+	const size_t lastDesc = (dev->lastTxDesc + nPackets) % SPW_TX_DESC_CNT;
+	bool wrapped = (lastDesc <= firstDesc);
+
+	for (size_t cnt = 0; cnt < nPackets; cnt++) {
+		(void)mutexLock(dev->txIrqLock);
+		while ((dev->txDescFree == 0) || dev->txWaited[dev->lastTxDesc]) {
+			/* Wait for free descriptor or for packet to be acknowledged */
+			(void)condWait(dev->cond, dev->txIrqLock, 0);
+		}
+		dev->txDescFree--;
+		(void)mutexUnlock(dev->txIrqLock);
+
+		volatile spw_txDesc_t *desc = &dev->txDesc[dev->lastTxDesc];
+
+		spw_txPacket_t packet;
+		buf += spw_txMsgToPacket(buf, &packet);
+		/* Interrupt after each packet transmitted */
+		desc->ctrl = (packet.flags & TX_DESC_USR_MSK) | TX_DESC_IE;
+		if (dev->lastTxDesc == SPW_TX_DESC_CNT - 1) {
+			/* Wrap around */
+			desc->ctrl |= TX_DESC_WR;
+		}
+		desc->hdrAddr = (void *)va2pa((void *)packet.hdr);
+		desc->packetLen = packet.dataLen;
+		desc->dataAddr = (void *)va2pa((void *)packet.data);
+
+		/* Everything is set up, enable descriptor */
+		desc->ctrl |= TX_DESC_EN;
+
+		dev->txWaited[dev->lastTxDesc] = true;
+
+		/* Start transmission */
+		dev->vbase[DMA_CTRL] |= DMA_CTRL_TE;
+
+		dev->lastTxDesc = (dev->lastTxDesc + 1) % SPW_TX_DESC_CNT;
+	}
+
+	TRACE("Packets set up");
+
+	/* Wait for transmission to finish */
+	while ((firstDesc <= lastDesc) || (wrapped && (firstDesc > lastDesc))) {
```

Comment:
```suggestion
	while ((firstDesc <= lastDesc) || wrapped) {
```
`firstDesc > lastDesc` is always `true`.